source/common/http/conn_manager_impl.cc

```diff
@@ -909,6 +909,40 @@ bool ConnectionManagerImpl::ActiveStream::validateHeaders() {
   return true;
 }
 
+bool ConnectionManagerImpl::ActiveStream::validateTrailers() {
+  if (header_validator_) {
+    auto validation_result = header_validator_->validateRequestTrailerMap(*request_trailers_);
+    if (!validation_result.ok()) {
+      Code response_code = Code::BadRequest;
+      absl::optional<Grpc::Status::GrpcStatus> grpc_status;
+      if (Grpc::Common::hasGrpcContentType(*request_headers_)) {
+        grpc_status = Grpc::Status::WellKnownGrpcStatus::Internal;
+      }
+
+      // H/2 codec was resetting requests that were rejected due to headers with underscores,
+      // instead of sending 400. Preserving this behavior for now.
+      // TODO(#24466): Make H/2 behavior consistent with H/1 and H/3.
+      if (validation_result.details() == UhvResponseCodeDetail::get().InvalidUnderscore &&
+          connection_manager_.codec_->protocol() == Protocol::Http2) {
+        filter_manager_.streamInfo().setResponseCodeDetails(validation_result.details());
+        resetStream();
+      } else {
+        if (connection_manager_.codec_->protocol() < Protocol::Http2) {
+          sendLocalReply(response_code, "", nullptr, grpc_status, validation_result.details());
+        } else {
+          filter_manager_.streamInfo().setResponseCodeDetails(validation_result.details());
+          resetStream();
```

Comment:
why not send a local response for H2 in this case?

Reply:
I have filed and Issue and left to TODO to harmonize the behavior. I will do it in a separate PR as it will affect a lot of tests, where the reset for H/2 and H/3 is expected.